### DISCUSSION TITLE
ABM Improvements Priority List

### DISCUSSION THREAD

**Discussion by xsaschako** [SciCompMod/memilio#1091] [General]
As we develop for the first paper, there are some things which need to be done afterward. I want to prioritize this so we can get going on that, while functionality should probably be done on a paper basis. I want to update this as we go along the developments of further papers.

High-Priority:
- Bug und performance low hanging fruits: There are a few.
- Setup of NPIs. We need a useful construct around this. This should include Masks, Capacity, Damping, Testing, Seasonality etc. This would also have the benefit, that we streamline the movement function.
- Setup of a simulation. This is similar to above, as above should make this 10x easier.
- Setup of a Grid Search
- Output: Vector of loggers and output

Mid-Priority:
- Useful OpenMP and MPI parallelization for these structures: Grid Search, Multiple Runs, Single Run. This needs to take into account for example the memory bounds of a cluster. 
- Infections. Right now we have a rate. With direct contacts, we'll have a probability. This needs to be explainable. Also which types we need and how we couple infectivity and symptomaticy
- Transition from Hospital to dead (comorbidities?)
- We need to think about what max contacts, capacity, and dampings exactly model and how they play into each other.
- Modularization of the ABM. This is to be discussed in detail what we want to achieve and to not overdo this.
- Location Types need to be discussed, e.g., two types of events, special events, multiple places to go to.
- Location types should have individual alphas as functions.
- Masks (migration rules need to wear mask according to rules, in trips we dont do that)

Low-Priority:
- adaptive step length, in the night 3 hours might be enough, on the day maybe 0.5 hours are justified
- Simplify certain unused aspects of the ABM. We would need to go through the whole ABM, but there are certainly things which are not used. As we are also memory bound for the grid search, deleting stuff should improve development time, interdependency, and data space take. Some things which came up during development: Cells, Capacity (not easy to work with), maybe just for Hospitals(?), Capacity Transmission risk, mask types, Activity Type in trips, trip origin.

**Comment by mknaranja** [comment]
Thanks for the extensive list. I have two items to be added and some comments.

Additions:

High- to mid-priority:
- Consider load balancing
- performance engineering
- sensitivity measures

Comments:

- Setup of NPIs. We need a useful construct around this. This should include Masks, Capacity, Damping, Testing, Seasonality etc. This would 
--> seasonality should not go into NPI.
- Location types should have individual alphas as functions.
--> should be explained
- Masks (migration rules need to wear mask according to rules, in trips we dont do that)
--> unclear
- adaptive step length, in the night 3 hours might be enough, on the day maybe 0.5 hours are justified
--> not sure this makes sense. If we have a clever execution of a time-step you don't do much work anyway over night so that it might make sense to keep the step size small(er)
